### PR TITLE
raftstore: remove `SnapshotDeleter`

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

### What problem does this PR solve?

The component `snap.rs` is too complex. Needs to be simplified.

### What is changed and how it works?

Remove `SnapshotDeleter` to simplify `raftstore/store/snap.rs`.
So we can do more improvements later.

### Check List

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note

### CODE REVIEW

File: components/raftstore/src/store/snap.rs [hunk 1]
```diff
@@ -1248,8 +1158,7 @@ where
 
     // Return all snapshots which is idle not being used.
     pub fn list_idle_snap(&self) -> io::Result<Vec<(SnapKey, bool)>> {
-        let core = self.core.rl();
-        let path = Path::new(&core.base);
+        let path = Path::new(&self.base);
```
Comment (BusyJay) [inline components/raftstore/src/store/snap.rs:1161]: Seems racy between generating and listing.

File: components/raftstore/src/store/snap.rs [hunk 2]
```diff
@@ -1338,34 +1260,17 @@ where
             };
         }
 
-        let (dir, snap_size) = {
```
Comment (BusyJay) [inline components/raftstore/src/store/snap.rs:1341]: I remember there was a deadlock problem.

File: components/raftstore/src/store/snap.rs [hunk 3]
```diff
@@ -1559,9 +1460,38 @@ where
         true
     }
 
+    pub fn retry_delete_snapshot(&self, key: &SnapKey, snap: &dyn GenericSnapshot) -> bool {
+        let d = time::Duration::from_millis(DELETE_RETRY_TIME_MILLIS);
+        for _ in 0..DELETE_RETRY_MAX_TIMES {
```
Comment (BusyJay) [inline components/raftstore/src/store/snap.rs:1465]: I think it may hold the lock too long.

File: components/raftstore/src/store/snap.rs [hunk 4]
```diff
@@ -493,36 +439,22 @@ impl Snap {
...
         key: &SnapKey,
-        size_track: Arc<AtomicU64>,
-        deleter: Box<dyn SnapshotDeleter>,
-        limiter: Limiter,
-        mgr: Option<Arc<DataKeyManager>>,
+        mgr: Arc<RwLock<SnapManagerCore>>,
     ) -> RaftStoreResult<Self> {
-        let mut s = Self::new(dir, key, size_track, true, true, deleter, limiter, mgr)?;
+        let mut s = Self::new(dir, key, true, true, mgr)?;
         s.init_for_building()?;
```
Comment (BusyJay) [inline components/raftstore/src/store/snap.rs:448]: There is no lock to protect building anymore, will it become a problem?

File: components/sst_importer/src/sst_importer.rs [hunk 5]
```diff
@@ -425,7 +428,7 @@ impl ImportDir {
         &self,
         meta: &SstMeta,
         engine: &E,
-        key_manager: Option<&Arc<DataKeyManager>>,
+        key_manager: Option<&DataKeyManager>,
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply components/sst_importer/src/sst_importer.rs:431]: It seems that no other threads would hold this object again.

File: components/raftstore/src/store/snap.rs (hunk 4 above)
Comment (hicqu) [reply components/raftstore/src/store/snap.rs:448]: Why? the lock scope is hot changed for now.

File: components/raftstore/src/store/snap.rs (hunk 2 above)
Comment (hicqu) [reply components/raftstore/src/store/snap.rs:1341]: What deadlock problem? All places requires `rl` except `regsiter`, `deregister` and `init`. So it seems ok.

File: components/sst_importer/src/sst_importer.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (hicqu) [reply components/sst_importer/src/sst_importer.rs:431]: And it doesn't need to be an `Arc`.

File: components/raftstore/src/store/snap.rs [hunk 6]
```diff
@@ -575,7 +520,7 @@ impl Snap {
                 write_digest: crc32fast::Hasher::new(),
             });
 
-            if let Some(mgr) = s.encryption_key_manager.as_ref() {
+            if let Some(mgr) = s.mgr.rl().encryption_key_manager() {
```
Comment (BusyJay) [inline components/raftstore/src/store/snap.rs:523]: Is it necessary to hold the lock while encrypting?

File: components/raftstore/src/store/snap.rs [hunk 7]
```diff
@@ -812,15 +747,18 @@ impl Snap {
                     cf_file.cf,
                     &begin_key,
                     &end_key,
-                    &self.limiter,
+                    &self.mgr.rl().limiter,
```
Comment (BusyJay) [inline components/raftstore/src/store/snap.rs:750]: Is it necessary to hold the read lock while building sst file?

File: components/raftstore/src/store/snap.rs [hunk 8]
```diff
@@ -493,36 +451,22 @@ impl Snap {
...
-            dir,
-            key,
-            size_track,
-            true,
-            false,
-            deleter,
-            Limiter::new(INFINITY),
-            mgr,
-        )?;
+        let mut s = Self::new(dir, key, true, false, mgr)?;
```
Comment (BusyJay) [inline components/raftstore/src/store/snap.rs:469]: Is it a problem that original limit is INFINITY?

File: components/raftstore/src/store/snap.rs [hunk 9]
```diff
@@ -575,7 +516,7 @@ impl Snap {
                 write_digest: crc32fast::Hasher::new(),
             });
 
-            if let Some(mgr) = s.encryption_key_manager.as_ref() {
+            if let Some(mgr) = s.mgr.encryption_key_manager.clone() {
```
Comment (BusyJay) [inline components/raftstore/src/store/snap.rs:519]: Seems `&s.mgr.encryption_key_manager` is enough.

File: components/raftstore/src/store/snap.rs [hunk 10]
```diff
@@ -659,8 +589,8 @@ impl Snap {
             cf_file.size = meta.get_size();
             cf_file.checksum = meta.get_checksum();
             if file_exists(&cf_file.path) {
-                let mgr = self.encryption_key_manager.as_ref().map(|t| t.as_ref());
-                let (_, size) = calc_checksum_and_size(&cf_file.path, mgr)?;
+                let mgr = self.mgr.encryption_key_manager.clone();
```
Comment (BusyJay) [inline components/raftstore/src/store/snap.rs:592]: Seems no need to clone.

File: components/raftstore/src/store/snap.rs [hunk 11]
```diff
@@ -711,14 +641,19 @@ impl Snap {
                     cf_file.checksum,
                 )?;
             }
-            let mgr = self.encryption_key_manager.as_ref().map(|x| x.as_ref());
-            check_file_size_and_checksum(&cf_file.path, cf_file.size, cf_file.checksum, mgr)?;
+            let mgr = self.mgr.encryption_key_manager.clone();
```
Comment (BusyJay) [inline components/raftstore/src/store/snap.rs:644]: Needless clone.

File: components/raftstore/src/store/snap.rs [hunk 12]
```diff
@@ -812,15 +746,15 @@ impl Snap {
                     cf_file.cf,
                     &begin_key,
                     &end_key,
-                    &self.limiter,
+                    &self.mgr.limiter.clone(),
```
Comment (BusyJay) [inline components/raftstore/src/store/snap.rs:749]: Needless clone.

File: components/raftstore/src/store/snap.rs (hunk 8 above)
Comment (hicqu) [reply components/raftstore/src/store/snap.rs:469]: I'm not sure why the original is INFINITY. I will keep the original value.

File: components/raftstore/src/store/snap.rs [hunk 13]
```diff
@@ -596,23 +538,12 @@ impl Snap {
...
-            dir,
-            key,
-            size_track,
-            false,
-            false,
-            deleter,
-            Limiter::new(INFINITY),
-            mgr,
-        )?;
+        let s = Self::new(dir, key, false, false, mgr)?;
```
Comment (BusyJay) [inline components/raftstore/src/store/snap.rs:546]: It's INFINITY in the past too.

File: components/raftstore/src/store/snap.rs (hunk 11 above)
Comment (BusyJay) [reply components/raftstore/src/store/snap.rs:645]: Seems it's missed.

**Pull request merged**